File: combinationsum.java
public class Solution {
    public ArrayList<ArrayList<Integer>> combinationSum(int[] candidates, int target) {
        // Start typing your Java solution below
        // DO NOT write main() function
        HashMap<Integer, ArrayList<ArrayList<Integer>>> numberToComb = 
            new HashMap<Integer, ArrayList<ArrayList<Integer>>>();
        ArrayList<Integer> cda = new ArrayList<Integer>();
        for(int i=0;i<candidates.length;i++){
            cda.add(candidates[i]);
        }
        Collections.sort(cda);
        HashSet<Integer> hs = new HashSet<Integer>(cda);
        
        
        for(int i=1;i<=target;i++){
            numberToComb.put(i, new ArrayList<ArrayList<Integer>>());
        }
        for(int i=1;i<=target;i++){
            if(hs.contains(i)){
                ArrayList<Integer> tmp = new ArrayList<Integer>();
                tmp.add(i);
                numberToComb.get(i).add(tmp);      
            }
            if(i==target){
                break;
            }
            for(Integer k : cda){
                if( k+i<=target ){
                    for(ArrayList<Integer> ar : numberToComb.get(i)){
                        if(k<ar.get(ar.size()-1)){
                            continue;
                        }
                        ArrayList<Integer> arnew = new ArrayList<Integer>(ar);
                        arnew.add(k);
                        
                        numberToComb.get(k+i).add(arnew);
                    }
                    
                }  
            }
            
        }
        return numberToComb.get(target);
    }
}

File: cpp/CombinationSum.cpp
class Solution {
public:
    vector<vector<int> > combinationSum(vector<int> &candidates, int target) {
        if(candidates.size()==0) {
            return vector<vector<int> >();
        }
        sort(candidates.begin(), candidates.end());
        return combinationSum(candidates, (int)candidates.size()-1, target);
    }
    
    vector<vector<int> > combinationSum(vector<int> &sortedCandidates, int end, int target) {
       if(end==-1) {
           if(target==0) {
                vector<int> v;
                vector<vector<int> > res;
                res.push_back(v);
                return res;
           } else {
               return vector<vector<int> >();
           }
       }
       vector<vector<int> > cWithoutLast = combinationSum(sortedCandidates, end-1, target);
       if(target < sortedCandidates[end]) {
           return cWithoutLast;
       }
       target -= sortedCandidates[end];
       vector<vector<int> > cWithLast = combinationSum(sortedCandidates, end, target);
       for(int i=0;i<cWithLast.size();i++) {
           cWithLast[i].push_back(sortedCandidates[end]);
       }
       cWithLast.insert(cWithLast.end(), cWithoutLast.begin(), cWithoutLast.end());
       return cWithLast;
    }
};
